addon/content/reducer.js

```diff
@@ -107,7 +128,167 @@ const attachmentActions = {
+// determine if this is worth sharing via a shared module with the background
+// scripts, or if it doesn't need it.
+const kScrollUnreadOrLast = 0;
+const kScrollSelected = 1;
+
+async function setupConversationInTab(params, isInTab) {
+  let scrollMode = params.get("scrollMode");
+  if (scrollMode) {
+    scrollMode = parseInt(scrollMode);
+  } else {
+    scrollMode = kScrollUnreadOrLast;
+  }
+  // If we start up Thunderbird with a saved conversation tab, then we
+  // have no selected message. Fallback to the usual mode.
+  if (
+    scrollMode == kScrollSelected &&
+    !topMail3Pane(window).gFolderDisplay.selectedMessage
+  ) {
+    scrollMode = kScrollUnreadOrLast;
+  }
+
+  if (window.frameElement) {
+    window.frameElement.setAttribute("tooltip", "aHTMLTooltip");
+  }
+  // let willExpand = parseInt(params.get("willExpand"));
+  const msgUrls = params.get("urls").split(",");
+  const msgIds = [];
+  for (const url of msgUrls) {
+    const id = await browser.conversations.getMessageIdForUri(url);
+    if (id) {
+      msgIds.push(id);
+    }
+  }
+  // It might happen that there are no messages left...
+  if (!msgIds.length) {
+    document.getElementById(
+      "messageList"
+    ).textContent = browser.i18n.getMessage(
+      "message.movedOrDeletedConversation"
+    );
+  } else {
+    window.Conversations = {
+      currentConversation: null,
+      counter: 0,
+    };
+    let freshConversation = new Conversation(
+      window,
+      // TODO: This should really become ids at some stage, but we need to
+      // teach Conversation how to handle those.
+      msgUrls,
+      scrollMode,
+      ++Conversations.counter,
+      isInTab
+    );
+    let browserFrame = window.frameElement;
+    // Because Thunderbird still hasn't fixed that...
+    if (browserFrame) {
+      browserFrame.setAttribute("context", "mailContext");
+    }
+
+    freshConversation.outputInto(window, async function(aConversation) {
+      // This is a stripped-down version of what's in monkeypatch.js,
+      //  make sure the two are in sync!
+      Conversations.currentConversation = aConversation;
+      aConversation.completed = true;
+      // TODO: Re-enable this.
+      // registerQuickReply();
+      // That's why we saved it before...
+      // newComposeSessionByDraftIf();
+      // TODO: expandQuickReply isn't defined anywhere. Should it be?
+      // if (willExpand)
+      //   expandQuickReply();
+      // Create a new rule that will override the default rule, so that
+      // the expanded quick reply is twice higher.
+      document.body.classList.add("inTab");
+      // Do this now so as to not defeat the whole expand/collapse
+      // logic.
+      if (
+        await browser.conversations.getCorePref(
+          "mailnews.mark_message_read.auto"
+        )
+      ) {
+        setTimeout(function() {
+          for (const id of msgIds) {
+            browser.messages.update(id, { read: true }).catch(console.error);
```

Comment:
It may be overkill, but should we check *again* that `mark_message_read.auto` is true in case they changed the setting while the timeout was still waiting?

Reply:
I think it is one of those things that I'd consider highly unlikely, and if they have changed the setting just after selecting a message, then they should be expecting it to get marked as read anyway.